[GlobalISel] Refactor extractParts()

chuongg3 opened the pull request:
Moved extractParts() and extractVectorParts() from LegalizerHelper
to Utils to be able to use it in different passes.

extractParts() will also try to use unmerge when doing irregular
splits where possible, falling back to extract elements when not.

ChuongGoh-arm committed:
[GlobalISel] Pre-commit for Refactor extractParts()

ChuongGoh-arm committed:
[GlobalISel] Refactor extractParts()

Moved extractParts() and extractVectorParts() from LegalizerHelper
to Utils to be able to use it in different passes.

aemerson reviewed (commented):
It's hard to see the code changes when the function is moved. Can you split this PR into 2 commits, one moving and one extending the functionality?

ChuongGoh-arm committed:
[GlobalISel] Refactor extractParts()

Try to use unmerge when doing irregular splits where
possible, falling back to extract elements when not.

ChuongGoh-arm committed:
fixup! [GlobalISel] Refactor extractParts()

aemerson commented on llvm/lib/CodeGen/GlobalISel/Utils.cpp:
```diff
@@ -506,6 +506,47 @@ bool llvm::extractParts(Register Reg, LLT RegTy, LLT MainTy, LLT &LeftoverTy,
     return true;
   }
 
+  // Try to use unmerge for irregular vector split where possible
```
I think a small MIR example would be helpful for those reading this code.

aemerson commented on llvm/test/CodeGen/AArch64/GlobalISel/legalize-shuffle-vector.mir:
```diff
@@ -460,12 +460,14 @@ body:             |
-    ; CHECK-NEXT: [[BUILD_VECTOR2:%[0-9]+]]:_(<4 x s32>) = G_BUILD_VECTOR [[EVEC]](s32), [[EVEC1]](s32), [[EVEC2]](s32), [[EVEC3]](s32)
-    ; CHECK-NEXT: [[BUILD_VECTOR3:%[0-9]+]]:_(<2 x s32>) = G_BUILD_VECTOR [[EVEC4]](s32), [[EVEC5]](s32)
-    ; CHECK-NEXT: G_STORE [[BUILD_VECTOR2]](<4 x s32>), [[COPY8]](p0) :: (store (<4 x s32>), align 32)
+    ; CHECK-NEXT: [[BUILD_VECTOR2:%[0-9]+]]:_(<2 x s32>) = G_BUILD_VECTOR [[EVEC]](s32), [[EVEC1]](s32)
+    ; CHECK-NEXT: [[BUILD_VECTOR3:%[0-9]+]]:_(<2 x s32>) = G_BUILD_VECTOR [[EVEC2]](s32), [[EVEC3]](s32)
+    ; CHECK-NEXT: [[BUILD_VECTOR4:%[0-9]+]]:_(<2 x s32>) = G_BUILD_VECTOR [[EVEC4]](s32), [[EVEC5]](s32)
+    ; CHECK-NEXT: [[CONCAT_VECTORS:%[0-9]+]]:_(<4 x s32>) = G_CONCAT_VECTORS [[BUILD_VECTOR2]](<2 x s32>), [[BUILD_VECTOR3]](<2 x s32>)
+    ; CHECK-NEXT: G_STORE [[CONCAT_VECTORS]](<4 x s32>), [[COPY8]](p0) :: (store (<4 x s32>), align 32)
```
This seems like a regression? Unless we have a combine that simplifies the 2 x build_vector + concat_vector into a single wide build vector. What do you think?

chuongg3 replied:
@davemgreen and I have been looking at this, to see if there is anywhere where it might cause problems. It looks like it can come up in some cases, but it usually limited to odd sized vectors. It looks like it would be a good optimization to add - I will add it to the list of things to do

aemerson replied:
I'm ok with this going in if that will be addressed later then.

dzhidzhoev commented on llvm/include/llvm/CodeGen/GlobalISel/Utils.h:
```diff
@@ -247,6 +248,24 @@ MachineInstr *getDefIgnoringCopies(Register Reg,
+                  MachineIRBuilder &MIRBuilder, MachineRegisterInfo &MRI);
+
+/// Version which handles irregular sub-vector splits.
+void extractVectorParts(Register Reg, unsigned NumElst,
```
nit: NumElst -> NumElts

qcolombet commented:
> Moved extractParts() and extractVectorParts() from LegalizerHelper to Utils to be able to use it in different passes.
>

Could you elaborate on which passes may want to use this functionality while not using the `LegalizerHelper` directly?

aemerson commented:
> > Moved extractParts() and extractVectorParts() from LegalizerHelper to Utils to be able to use it in different passes.
> 
> Could you elaborate on which passes may want to use this functionality while not using the `LegalizerHelper` directly?

These are intended to be used in a combiner here: https://github.com/llvm/llvm-project/pull/75832

davemgreen reviewed (approved):
This LGTM, to go along with #75832, so long as the remaining comments are addressed. It looks like a useful routine to be able to use elsewhere.

ChuongGoh-arm committed:
fixup! fixup! [GlobalISel] Refactor extractParts()

s-barannikov commented on llvm/include/llvm/CodeGen/GlobalISel/Utils.h:
```diff
@@ -247,6 +248,24 @@ MachineInstr *getDefIgnoringCopies(Register Reg,
+/// registers created are appended to Ops, starting at bit 0 of Reg.
+void extractParts(Register Reg, LLT Ty, int NumParts,
+                  SmallVectorImpl<Register> &VRegs,
+                  MachineIRBuilder &MIRBuilder, MachineRegisterInfo &MRI);
```
nit: There is in-class `MachineIRBuidler::getMRI()`, no need to pass `MRI`.
